op-chain-ops/cmd/celo-migrate/continuity.go

```diff
@@ -0,0 +1,153 @@
+		td:            r.tds[i],
+	}, nil
+}
+
+func (r *RLPBlockRange) DropFirst() {
+	r.start = r.start + 1
+	r.hashes = r.hashes[1:]
+	r.headers = r.headers[1:]
+	r.bodies = r.bodies[1:]
+	r.receipts = r.receipts[1:]
+	r.tds = r.tds[1:]
+}
+
+// CheckContinuity checks if the block data in the range is continuous
+// by comparing the header number and parent hash of each block with the previous block,
+// and by checking if the number of elements retrieved from each table is the same.
+// It takes in the last element in the preceding range, and returns the last element
+// in the current range so that continuity can be checked across ranges.
+func (r *RLPBlockRange) CheckContinuity(prevElement *RLPBlockElement, expectedLength uint64) (*RLPBlockElement, error) {
+	log.Info("Checking data continuity for block range",
+		"start", r.start,
+		"end", r.start+expectedLength-1,
+		"count", expectedLength,
+		"prevElement", func() interface{} {
+			if prevElement != nil {
+				return prevElement.Header().Number.Uint64()
+			}
+			return "nil"
+		}(),
+	)
+
+	if err := r.CheckLengths(expectedLength); err != nil {
+		return nil, err
+	}
+	for i := range r.hashes {
+		currElement, err := r.Element(uint64(i))
+		if err != nil {
+			return nil, err
+		}
+		if currElement.Header().Number.Uint64() != r.start+uint64(i) {
+			return nil, fmt.Errorf("decoded header number mismatch indicating a gap in block numbers: expected %d, actual %d", r.start+uint64(i), currElement.Header().Number.Uint64())
+		}
+		if prevElement != nil {
+			log.Debug("Checking continuity", "block", currElement.Header().Number.Uint64(), "prev", prevElement.Header().Number.Uint64())
+			if err := currElement.Follows(prevElement); err != nil {
+				return nil, err
+			}
+		}
+		prevElement = currElement
+	}
+	return prevElement, nil
+}
+
+// CheckLengths makes sure the number of elements retrieved from each table is the same
+func (r *RLPBlockRange) CheckLengths(expectedLength uint64) error {
+	var err error
+	if uint64(len(r.hashes)) != expectedLength {
+		err = fmt.Errorf("Expected count mismatch in block range hashes: expected %d, actual %d", expectedLength, len(r.hashes))
+	}
+	if uint64(len(r.bodies)) != expectedLength {
+		err = errors.Join(err, fmt.Errorf("Expected count mismatch in block range bodies: expected %d, actual %d", expectedLength, len(r.bodies)))
+	}
+	if uint64(len(r.headers)) != expectedLength {
+		err = errors.Join(err, fmt.Errorf("Expected count mismatch in block range headers: expected %d, actual %d", expectedLength, len(r.headers)))
+	}
+	if uint64(len(r.receipts)) != expectedLength {
+		err = errors.Join(err, fmt.Errorf("Expected count mismatch in block range receipts: expected %d, actual %d", expectedLength, len(r.receipts)))
+	}
+	if uint64(len(r.tds)) != expectedLength {
+		err = errors.Join(err, fmt.Errorf("Expected count mismatch in block range total difficulties: expected %d, actual %d", expectedLength, len(r.tds)))
+	}
+	return err
+}
```

Comment:
This error message is slightly confusing.

If read one way it can make it sound like  a "count mismatch" was expected. I'd suggest something like:

```
Unexpected length for block range.....
```